README.md
# MapComplete

> Let a thousand flowers bloom

MapComplete is an OpenStreetMap viewer and editor. It shows map features on a certain topic, and allows to see, edit and add new features to the map.
It can be seen as a webversion [crossover of StreetComplete and MapContrib](Docs/MapComplete_vs_other_editors.md). It tries to be just as easy to use as StreetComplete, but it allows to focus on one single theme per instance (e.g. nature, bicycle infrastructure, ...)

The design goals of MapComplete are to be:

- Easy to use, both on web and on mobile
- Easy to deploy (by not having a backend)
- Easy to set up a custom theme
- Easy to fall down the rabbit hole of OSM

The basic functionality is to download some map features from Overpass and then ask certain questions. An answer is sent back to directly to OpenStreetMap.

Furthermore, it shows images present in the `image` tag or, if a `wikidata` or `wikimedia_commons`-tag is present, it follows those to get these images too.

An explicit non-goal of MapComplete is to modify geometries of ways. Although adding a point to a way or splitting a way in two parts might be added one day.

# Creating your own theme

It is possible to quickly make and distribute your own theme - [please read the documentation on how to do this](Docs/Making_Your_Own_Theme.md).

## Examples

- [Buurtnatuur.be](http://buurtnatuur.be), developed for the Belgian [Green party](https://www.groen.be/). They also funded the initial development!
- [Cyclofix](https://pietervdvn.github.io/MapComplete/index.html?layout=cyclofix), further development on [Open Summer of Code](https://summerofcode.be/) funded by [Brussels Mobility](https://mobilite-mobiliteit.brussels/en). Landing page at https://cyclofix.osm.be/
- [Bookcases](https://pietervdvn.github.io/MapComplete/index.html?quests=bookcases#element) cause I like to collect them.
- [Map of Maps](https://pietervdvn.github.io/MapComplete/index.html?layout=maps&z=14&lat=50.650&lon=4.2668#element), after a tweet

There are plenty more. Discover them in the app.

### Statistics

To see statistics, consult [OsmCha](https://osmcha.org/?filters=%7B%22comment%22%3A%5B%7B%22label%22%3A%22%23mapcomplete%22%2C%22value%22%3A%22%23mapcomplete%22%7D%5D%2C%22date__gte%22%3A%5B%7B%22label%22%3A%222020-07-05%22%2C%22value%22%3A%222020-07-05%22%7D%5D%7D) or the [analytics page](https://pietervdvn.goatcounter.com/)

## User journey

MapComplete is set up to lure people into OpenStreetMap and to teach them while they are on the go, step by step.

A typical user journey would be:

0. Oh, this is a cool map of _my specific interest_! There is a lot of data already...

   - The user might discover the explanation about OSM in the second tab
   - The user might share the map and/or embed it in the third tab
   - The user might discover the other themes in the last tab

1. The user clicks that big tempting button 'login' in order to answer questions - there's enough of these login buttons... The user creates an account.

2. The user answers a question! Hooray! The user transformed into a **contributor** now.

   - When at least one question is answered (aka: having one changeset on OSM), adding a new point is unlocked

3. The user adds a new POI somewhere

   - Note that _all messages_ must be read before being able to add a point.
   - In other words, sending a message to a misbehaving MapComplete user acts as having a **zero-day-block**. This is added deliberately to make sure new users _have_ to read feedback from the community.

4. At 50 changesets, the [personal layout](https://pietervdvn.github.io/MapComplete/personal.html) is advertised. The personal theme is a theme where contributors can pick layers from all the official themes. Note that the personal theme is always available.

5. At 200 changesets, the tags become visible when answering questions and when adding a new point from a preset. This is to give more control to power users and to teach new users the tagging scheme

6. At 250 changesets, the tags get linked to the wiki

7. At 500 changesets, I expect contributors to be power users and to be comfortable with tagging scheme and such. The custom theme generator is unlocked.

## License

GPLv3.0 + recommended pingback.

I love it to see where the project ends up. You are free to reuse the software (under GPL) but, when you have made your own change and are using it, I would like to know about it. Drop me a line, give a pingback in the issues,...

## Dev

To develop or deploy a version of MapComplete, have a look [to the guide](Docs/Development_deployment.md).

## Translating MapComplete

The core strings and builtin themes of MapComplete are translated on [Hosted Weblate](https://hosted.weblate.org/projects/mapcomplete/core/).
You can easily make an account and start translating in their web-environment - no installation required.

[![Translation status](https://hosted.weblate.org/widgets/mapcomplete/-/multi-blue.svg)](https://hosted.weblate.org/engage/mapcomplete/)
[![Translation status](https://hosted.weblate.org/widgets/mapcomplete/-/multi-blue.svg)](https://hosted.weblate.org/engage/mapcomplete/)

## Architecture

### High-level overview

The website is purely static. This means that there is no database here, nor one is needed as all the data is kept in OpenStreetMap, Wikimedia (for images), Imgur. Settings are saved in the preferences-space of the OSM-website, amended by some local-storage if the user is not logged-in.

When viewing, the data is loaded from overpass. The data is then converted (in the browser) to geojson, which is rendered by Leaflet.

When a map feature is clicked, a popup shows the information, images and questions that are relevant for that object.
The answers given by the user are sent (after a few seconds) to OpenStreetMap directly - if the user is logged in. If not logged in, the user is prompted to do so.

The UI-event-source is a class where the entire system is built upon, it acts as an observable object: another object can register for changes to update when needed.

### Searching images

Images are fetched from:

- The OSM `image`, `image:0`, `image:1`, ... tags
- The OSM `wikimedia_commons` tags
- If wikidata is present, the wikidata `P18` (image) claim and, if a commons link is present, the commons images

### Uploading images

Images are uploaded to Imgur, as their API was way easier to handle. The URL is written into the changes.

The idea is that once in a while, the images are transferred to wikipedia or that we hook up wikimedia directly (but I need some help in getting their API working).

### Uploading changes

In order to avoid lots of small changesets, a changeset is opened and kept open. The changeset number is saved into the users preferences on OSM.

Whenever a change is made -even adding a single tag - the change is uploaded into this changeset. If that fails, the changeset is probably closed and we open a new changeset.

Note that changesets are closed automatically after one hour of inactivity, so we don't have to worry about closing them.

# Documentation

All documentation can be found in [here](Docs/)

# Privacy

Privacy is important, we try to leak as little information as possible.
All major personal information is handled by OSM.
Geolocation is available on mobile only through the device's GPS location (so no geolocation is sent of to Google).

TODO: erase cookies of third party websites and API's

# Attribution and Copyright

The code is available under GPL; all map data comes from OpenStreetMap (both foreground and background maps).

Background layer selection: curated by https://github.com/osmlab/editor-layer-index

Icons are attributed in various 'license_info.json'-files and can be found in the app.


Docs/Release_Notes.md
# Release Notes

Some highlights of new releases.

## 0.7.0

- New theme: Hailhydrant by @GoWIN/Erwin Olario
- Tags now include metainformation by default - see [CalculatedTags.md]
- Add a membership-function in the calculated tags - useful to add 'part of walking route XXX' on a way
- Fix login when the token has been revoked
- First, experimental support for tiled geojson layers + scripts to generate those layers (see the speelplekken-theme)
- Add a global override
- Ton's of bug fixes


Utils.ts
import * as colors from './assets/colors.json';
import { TileRange } from './Models/TileRange';

export class Utils {
  /**
   * In the 'deploy'-step, some code needs to be run by ts-node.
   * However, ts-node crashes when it sees 'document'. When running from console, we flag this and disable all code where document is needed.
   * This is a workaround and yet another hack
   */
  public static runningFromConsole = false;
  public static readonly assets_path = './assets/svg/';
  public static externalDownloadFunction: (url: string) => Promise<any>;
  private static knownKeys = [
    'addExtraTags',
    'and',
    'calculatedTags',
    'changesetmessage',
    'clustering',
    'color',
    'condition',
    'customCss',
    'dashArray',
    'defaultBackgroundId',
    'description',
    'descriptionTail',
    'doNotDownload',
    'enableAddNewPoints',
    'enableBackgroundLayerSelection',
    'enableGeolocation',
    'enableLayers',
    'enableMoreQuests',
    'enableSearch',
    'enableShareScreen',
    'enableUserBadge',
    'freeform',
    'hideFromOverview',
    'hideInAnswer',
    'icon',
    'iconOverlays',
    'iconSize',
    'id',
    'if',
    'ifnot',
    'isShown',
    'key',
    'language',
    'layers',
    'lockLocation',
    'maintainer',
    'mappings',
    'maxzoom',
    'maxZoom',
    'minNeededElements',
    'minzoom',
    'multiAnswer',
    'name',
    'or',
    'osmTags',
    'passAllFeatures',
    'presets',
    'question',
    'render',
    'roaming',
    'roamingRenderings',
    'rotation',
    'shortDescription',
    'socialImage',
    'source',
    'startLat',
    'startLon',
    'startZoom',
    'tagRenderings',
    'tags',
    'then',
    'title',
    'titleIcons',
    'type',
    'version',
    'wayHandling',
    'widenFactor',
    'width'
  ];
  private static extraKeys = [
    'nl',
    'en',
    'fr',
    'de',
    'pt',
    'es',
    'name',
    'phone',
    'email',
    'amenity',
    'leisure',
    'highway',
    'building',
    'yes',
    'no',
    'true',
    'false'
  ];

  static EncodeXmlValue(str) {
    if (typeof str !== 'string') {
      str = '' + str;
    }

    return str.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;').replace(/'/g, '&apos;');
  }

  /**
   * Gives a clean float, or undefined if parsing fails
   * @param str
   */
  static asFloat(str): number {
    if (str) {
      const i = parseFloat(str);
      if (isNaN(i)) {
        return undefined;
      }
      return i;
    }
    return undefined;
  }

  public static Upper(str: string) {
    return str.substr(0, 1).toUpperCase() + str.substr(1);
  }

  public static TwoDigits(i: number) {
    if (i < 10) {
      return '0' + i;
    }
    return '' + i;
  }

  public static Round(i: number) {
    if (i < 0) {
      return '-' + Utils.Round(-i);
    }
    const j = '' + Math.floor(i * 10);
    if (j.length == 1) {
      return '0.' + j;
    }
    return j.substr(0, j.length - 1) + '.' + j.substr(j.length - 1, j.length);
  }

  public static Times(f: (i: number) => string, count: number): string {
    let res = '';
    for (let i = 0; i < count; i++) {
      res += f(i);
    }
    return res;
  }

  public static TimesT<T>(count: number, f: (i: number) => T): T[] {
    let res: T[] = [];
    for (let i = 0; i < count; i++) {
      res.push(f(i));
    }
    return res;
  }

  static DoEvery(millis: number, f: () => void) {
    if (Utils.runningFromConsole) {
      return;
    }
    window.setTimeout(function () {
      f();
      Utils.DoEvery(millis, f);
    }, millis);
  }

  public static NoNull<T>(array: T[]): T[] {
    const ls: T[] = [];
    for (const t of array) {
      if (t === undefined || t === null) {
        continue;
      }
      ls.push(t);
    }
    return ls;
  }

  public static NoEmpty(array: string[]): string[] {
    const ls: string[] = [];
    for (const t of array) {
      if (t === '') {
        continue;
      }
      ls.push(t);
    }
    return ls;
  }

  public static EllipsesAfter(str: string, l: number = 100) {
    if (str === undefined || str === null) {
      return undefined;
    }
    if (str.length <= l) {
      return str;
    }
    return str.substr(0, l - 3) + '...';
  }

  public static Dedup(arr: string[]): string[] {
    if (arr === undefined) {
      return undefined;
    }
    const newArr = [];
    for (const string of arr) {
      if (newArr.indexOf(string) < 0) {
        newArr.push(string);
      }
    }
    return newArr;
  }

  public static MergeTags(a: any, b: any) {
    const t = {};
    for (const k in a) {
      t[k] = a[k];
    }
    for (const k in b) {
      t[k] = b[k];
    }
    return t;
  }

  public static SplitFirst(a: string, sep: string): string[] {
    const index = a.indexOf(sep);
    if (index < 0) {
      return [a];
    }
    return [a.substr(0, index), a.substr(index + sep.length)];
  }

  public static SubstituteKeys(txt: string, tags: any) {
    for (const key in tags) {
      if (!tags.hasOwnProperty(key)) {
        continue;
      }
      txt = txt.replace(new RegExp('{' + key + '}', 'g'), tags[key]);
    }
    return txt;
  }

  public static LoadCustomCss(location: string) {
    const head = document.getElementsByTagName('head')[0];
    const link = document.createElement('link');
    link.id = 'customCss';
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = location;
    link.media = 'all';
    head.appendChild(link);
    console.log('Added custom layout ', location);
  }

  /**
   * Copies all key-value pairs of the source into the target.
   * If the key starts with a '+', the values of the list will be appended to the target instead of overwritten
   * @param source
   * @param target
   * @constructor
   */
  static Merge(source: any, target: any) {
    for (const key in source) {
      if (!source.hasOwnProperty(key)) {
        continue;
      }
      if (key.startsWith('+') || key.endsWith('+')) {
        const trimmedKey = key.replace('+', '');
        const sourceV = source[key];
        const targetV = target[trimmedKey] ?? [];

        let newList: any[];
        if (key.startsWith('+')) {
          newList = sourceV.concat(targetV);
        } else {
          newList = targetV.concat(sourceV);
        }

        target[trimmedKey] = newList;
        continue;
      }

      const sourceV = source[key];
      const targetV = target[key];
      if (sourceV?.length !== undefined && targetV?.length !== undefined && key.startsWith('+')) {
        target[key] = targetV.concat(sourceV);
      } else if (typeof sourceV === 'object') {
        if (targetV === undefined) {
          target[key] = sourceV;
        } else {
          Utils.Merge(sourceV, targetV);
        }
      } else {
        target[key] = sourceV;
      }
    }
    return target;
  }

  static getOrSetDefault<K, V>(dict: Map<K, V>, k: K, v: () => V) {
    let found = dict.get(k);
    if (found !== undefined) {
      return found;
    }
    dict.set(k, v());
    return dict.get(k);
  }

  /**
   * Calculates the tile bounds of the
   * @param z
   * @param x
   * @param y
   * @returns [[maxlat, minlon], [minlat, maxlon]]
   */
  static tile_bounds(z: number, x: number, y: number): [[number, number], [number, number]] {
    return [
      [Utils.tile2lat(y, z), Utils.tile2long(x, z)],
      [Utils.tile2lat(y + 1, z), Utils.tile2long(x + 1, z)]
    ];
  }

  /**
   * Return x, y of the tile containing (lat, lon) on the given zoom level
   */
  static embedded_tile(lat: number, lon: number, z: number): { x: number; y: number; z: number } {
    return { x: Utils.lon2tile(lon, z), y: Utils.lat2tile(lat, z), z: z };
  }

  static TileRangeBetween(zoomlevel: number, lat0: number, lon0: number, lat1: number, lon1: number): TileRange {
    const t0 = Utils.embedded_tile(lat0, lon0, zoomlevel);
    const t1 = Utils.embedded_tile(lat1, lon1, zoomlevel);

    const xstart = Math.min(t0.x, t1.x);
    const xend = Math.max(t0.x, t1.x);
    const ystart = Math.min(t0.y, t1.y);
    const yend = Math.max(t0.y, t1.y);
    const total = (1 + xend - xstart) * (1 + yend - ystart);

    return {
      xstart: xstart,
      xend: xend,
      ystart: ystart,
      yend: yend,
      total: total,
      zoomlevel: zoomlevel
    };
  }

  public static MinifyJSON(stringified: string): string {
    stringified = stringified.replace(/\|/g, '||');

    const keys = Utils.knownKeys.concat(Utils.extraKeys);
    for (let i = 0; i < keys.length; i++) {
      const knownKey = keys[i];
      let code = i;
      if (i >= 124) {
        code += 1; // Character 127 is our 'escape' character |
      }
      let replacement = '|' + String.fromCharCode(code);
      stringified = stringified.replace(new RegExp(`\"${knownKey}\":`, 'g'), replacement);
    }

    return stringified;
  }

  public static UnMinify(minified: string): string {
    if (minified === undefined || minified === null) {
      return undefined;
    }

    const parts = minified.split('|');
    let result = parts.shift();
    const keys = Utils.knownKeys.concat(Utils.extraKeys);

    for (const part of parts) {
      if (part == '') {
        // Empty string => this was a || originally
        result += '|';
        continue;
      }
      const i = part.charCodeAt(0);
      result += '"' + keys[i] + '":' + part.substring(1);
    }

    return result;
  }

  public static MapRange<T>(tileRange: TileRange, f: (x: number, y: number) => T): T[] {
    const result: T[] = [];
    for (let x = tileRange.xstart; x <= tileRange.xend; x++) {
      for (let y = tileRange.ystart; y <= tileRange.yend; y++) {
        const t = f(x, y);
        result.push(t);
      }
    }
    return result;
  }

  public static downloadJson(url: string): Promise<any> {
    if (this.externalDownloadFunction !== undefined) {
      return this.externalDownloadFunction(url);
    }

    return new Promise((resolve, reject) => {
      try {
        const xhr = new XMLHttpRequest();
        xhr.onload = () => {
          if (xhr.status == 200) {
            try {
              resolve(JSON.parse(xhr.response));
            } catch (e) {
              reject('Not a valid json: ' + xhr.response);
            }
          } else {
            reject(xhr.statusText);
          }
        };
        xhr.open('GET', url);
        xhr.setRequestHeader('accept', 'application/json');
        xhr.send();
      } catch (e) {
        reject(e);
      }
    });
  }

  /**
   * Triggers a 'download file' popup which will download the contents
   * @param contents
   * @param fileName
   */
  public static offerContentsAsDownloadableFile(contents: string, fileName: string = 'download.txt') {
    const element = document.createElement('a');
    const file = new Blob([contents], { type: 'text/plain' });
    element.href = URL.createObjectURL(file);
    element.download = fileName;
    document.body.appendChild(element); // Required for this to work in FireFox
    element.click();
  }

  public static ColourNameToHex(color: string): string {
    return colors[color.toLowerCase()] ?? color;
  }

  public static HexToColourName(hex: string): string {
    hex = hex.toLowerCase();
    if (!hex.startsWith('#')) {
      return hex;
    }
    const c = Utils.color(hex);

    let smallestDiff = Number.MAX_VALUE;
    let bestColor = undefined;
    for (const color in colors) {
      if (!colors.hasOwnProperty(color)) {
        continue;
      }
      const foundhex = colors[color];
      if (typeof foundhex !== 'string') {
        continue;
      }
      if (foundhex === hex) {
        return color;
      }
      const diff = this.colorDiff(Utils.color(foundhex), c);
      if (diff > 50) {
        continue;
      }
      if (diff < smallestDiff) {
        smallestDiff = diff;
        bestColor = color;
      }
    }
    return bestColor ?? hex;
  }

  private static tile2long(x, z) {
    return (x / Math.pow(2, z)) * 360 - 180;
  }

  private static tile2lat(y, z) {
    const n = Math.PI - (2 * Math.PI * y) / Math.pow(2, z);
    return (180 / Math.PI) * Math.atan(0.5 * (Math.exp(n) - Math.exp(-n)));
  }

  private static lon2tile(lon, zoom) {
    return Math.floor(((lon + 180) / 360) * Math.pow(2, zoom));
  }

  private static lat2tile(lat, zoom) {
    return Math.floor(
      ((1 - Math.log(Math.tan((lat * Math.PI) / 180) + 1 / Math.cos((lat * Math.PI) / 180)) / Math.PI) / 2) * Math.pow(2, zoom)
    );
  }

  private static colorDiff(c0: { r: number; g: number; b: number }, c1: { r: number; g: number; b: number }) {
    return Math.abs(c0.r - c1.r) + Math.abs(c0.g - c1.g) + Math.abs(c0.b - c1.b);
  }

  private static color(hex: string): { r: number; g: number; b: number } {
    if (hex.startsWith == undefined) {
      console.trace('WUT?', hex);
      throw 'wut?';
    }
    if (!hex.startsWith('#')) {
      return undefined;
    }
    if (hex.length === 4) {
      return {
        r: parseInt(hex.substr(1, 1), 16),
        g: parseInt(hex.substr(2, 1), 16),
        b: parseInt(hex.substr(3, 1), 16)
      };
    }

    return {
      r: parseInt(hex.substr(1, 2), 16),
      g: parseInt(hex.substr(3, 2), 16),
      b: parseInt(hex.substr(5, 2), 16)
    };
  }
}
